src/hash/hasher.cr

```diff
@@ -0,0 +1,225 @@
+require "crystal/system/random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+# ```
+# class MyHasher
+#   # Value should implement commutative `+` for `Hash#hash(hasher)`
+#   alias Value
+#
+#   # must be implemented to mix sizes of collections, and pointers (object_id)
+#   def raw(v : Int::Primitive)
+#     # mutate
+#     nil
+#   end
+#
+#   # must be implemented for Hash#hash
+#   def raw(v : Value)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(b : Bytes)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(n : Nil)
+#     # mutate
+#     nil
+#   end
+#
+#   def <<(v)
+#     # v.hash will return hasher
+#     # if hasher is a struct, then it will be copy
+#     copy_from v.hash(self)
+#     nil
+#   end
+#
+#   # digest returns hashsum for current state without state mutation
+#   def digest : Value
+#   end
+#
+#   # should be implemented for `Hash#hash(hasher)`
+#   def clone
+#     copy_of_current_state
+#   end
+# end
+# ```
+
+class Hash
+  # Hasher used as standard hasher in `Object#hash`
+  # It have to provide defense against HashDos, and to be reasonably fast.
+  #
+  # To protect against HashDos, it is seeded with secure random, and have
+  # permutation that hard to forge without knowing seed and seeing hash digest.
+  struct Hasher
```

Comment:
I'm not "well known expert", and I cannot proof anything with mathematics.

It is not "funny_hash", cause llvm sucks :-( . I made special hash algorithm for Crystal to workaround all llvm optimizer issues ;-( Initially I wanted to put `lucky777` (also mine algorithm, but heavilly inspired by well known Marvin32), but llvm failed to produce rot instructions when optimized that.

I've made this hash with all my experience to avoid trivial seed independent collisions. So it is *hard* to break without knowing seed and/or result, but *not impossible*.
I tried to make it hard enough so HashDoS will be as difficult as pure DoS, so attacker will choose to perform DoS directly than spend several days to prepare HashDoS.